src/tiled/wangpainter.cpp

```diff
@@ -0,0 +1,413 @@
+                    break;
+                case WangId::TopLeft:
+                    return WangId::Top;
+                    break;
+                case WangId::TopRight:
+                    return WangId::Right;
+                    break;
+                default:
+                    return initialDirection;
+                    break;
+            }
+            break;
+        
+        case BrushMode::PaintEdgeAndCorner:
+            switch (initialDirection) {
+                case WangId::BottomRight:
+                    return WangId::TopLeft;
+                    break;
+                case WangId::BottomLeft:
+                    return WangId::TopLeft;
+                    break;
+                case WangId::TopRight:
+                    return WangId::TopLeft;
+                    break;
+                default:
+                    return initialDirection;
+                    break;
+            }
+            break;
+    }
+}
+
+void WangPainter::setTerrain(WangFiller::FillRegion &fill, MapDocument *mapDocument, int color, QPoint pos, WangId::Index directionToGenerate, bool useTileMode) {
+    setColor(color);
+    WangId::Index direction = getDesiredDirection(directionToGenerate);
+    generateTerrainAt(mapDocument, fill, mCurrentColor, pos, direction, useTileMode);
+}
+
+void WangPainter::setTerrain(MapDocument *mapDocument, int color, QPoint pos, WangId::Index directionToGenerate, bool useTileMode) {
+    setTerrain(mCurrentFill, mapDocument, color, pos, directionToGenerate, useTileMode);
+}
+
+void WangPainter::clear() {
+    WangFiller::FillRegion newFill;
+    mCurrentFill = newFill;
+}
+
+void WangPainter::commit(MapDocument *mapDocument, TileLayer *tileLayer) {
+    SharedTileLayer stamp = SharedTileLayer::create(QString(), 0, 0, 0, 0);
+    WangFiller wangFiller{*mWangSet, mapDocument->renderer()};
+    wangFiller.setCorrectionsEnabled(true);
+
+    wangFiller.fillRegion(*stamp, *tileLayer, mCurrentFill.region, mCurrentFill.grid);
+
+    QRegion brushRegion = stamp->region([](const Cell &cell)
+                                            { return cell.checked(); });
+    brushRegion.translate(tileLayer->position());
+    QRect brushRect = brushRegion.boundingRect();
+    stamp->setPosition(brushRect.topLeft());
+    stamp->resize(brushRect.size(), -brushRect.topLeft());
+
+    for (int j = 0; j < stamp->height(); ++j) {
+        for (int i = 0; i < stamp->width(); ++i) {
+            Cell cell = stamp->cellAt(i, j);
+            if (cell.tileset() == nullptr)
+            {
+                continue;
+            }
+            if (cell.tile() == nullptr)
+            {
+                continue;
+            }
+            
+            tileLayer->setCell(stamp->x() + i, stamp->y() + j, cell);
+        }
+    }
+    clear();
+}
+
+void WangPainter::generateTerrainAt(MapDocument *mapDocument, WangFiller::FillRegion &fill, int color, QPoint pos, WangId::Index direction, bool useTileMode) {
```

Comment:
This function used to be `WangBrush::updateBrushAt`, which was moved here because it's core to making sure the `FillRegion` is populated correctly. The version now in `wangpainter` uses less internal state by putting more of that into the function's parameters, but the logic should be unchanged.